plugin/db/util/driverutil.go

```diff
@@ -216,36 +203,40 @@ func beginMigration(ctx context.Context, executor MigrationExecutor, m *db.Migra
 		}
 	}
 
-	// Check if there is any higher version already been applied
-	if version, err := executor.CheckOutOfOrderVersion(ctx, tx, m.Namespace, m.Source, m.Version); err != nil {
+	sqldb, err := executor.GetDbConnection(ctx, bytebaseDatabase)
+	if err != nil {
+		return -1, err
+	}
+	tx, err := sqldb.BeginTx(ctx, nil)
+	if err != nil {
 		return -1, err
-	} else if version != nil && len(*version) > 0 {
-		// Clickhouse will always return non-nil version with empty string.
-		return -1, common.Errorf(common.MigrationOutOfOrder, fmt.Errorf("database %q has already applied version %s which is higher than %s", m.Database, *version, m.Version))
-	}
-
-	// If the migration engine is VCS and type is not baseline and is not branch, then we can only proceed if there is existing baseline
-	// This check is also wrapped in transaction to avoid edge case where two baselinings are running concurrently.
-	if m.Source == db.VCS && m.Type != db.Baseline && m.Type != db.Branch {
-		if hasBaseline, err := executor.FindBaseline(ctx, tx, m.Namespace); err != nil {
-			return -1, err
-		} else if !hasBaseline {
-			return -1, common.Errorf(common.MigrationBaselineMissing, fmt.Errorf("%s has not created migration baseline yet", m.Database))
-		}
 	}
+	defer tx.Rollback()
 
-	// VCS based SQL migration requires existing baselining
-	requireBaseline := m.Source == db.VCS && m.Type == db.Migrate
-	sequence, err := executor.FindNextSequence(ctx, tx, m.Namespace, requireBaseline)
+	latestSequence, err := executor.FindLatestSequence(ctx, tx, m.Namespace)
 	if err != nil {
 		return -1, err
 	}
 
-	// Phase 2 - Record migration history as PENDING
+	// Check if there is any higher version already been applied.
+	if latestSequence > 0 && m.Type != db.Baseline && m.Type != db.Branch {
+		if list, err := executor.FindMigrationHistoryList(ctx, &db.MigrationHistoryFind{
+			Database: &m.Namespace,
+			Sequence: &latestSequence,
+		}); err != nil {
+			return -1, fmt.Errorf("Find latest migration error: %q", err)
+		} else if len(list) > 0 {
+			if list[0].Version > m.Version {
```

Comment:
This is not the same comparing with the old query

"		SELECT MIN(version) FROM bytebase.migration_history
		WHERE namespace = ? AND source = ? AND version > ?"
		
With the new implementation, we assume the record with the largest sequence has the latest version, which may not be the case if the migration is already out of order.

Reply:
Also, suppose we use the current logic, the comparison should be list[0].Version >= m.Version ?

Reply:
Changed to ">=".

With the new algorithm, out-of-ordering is expected to happen when users do a baseline with a lower version string.

Reply:
The new logic assumes the largest sequence has the largest version. This assumption is only true if the existing records are NOT out of order. We don't have database constraints to guarantee that. 

The ideal logic is to make sure the new version string is larger than all the previous migration records up to the last baseline.

Reply:
Correct, the condition "largestSequence > 0 && m.Type != db.Baseline && m.Type != db.Branch" gives us that.

"new version string is larger than all the previous migration records up to the last baseline" is similar to an overview (recursion), while the current condition is a local decision at the current step (recursive step).

Reply:
Let me give an example,

Sequence,  Version, Type
1,  v1.1.2, Migrate
2,  v1.1.0, Migrate

The user specifies v1.1.1 in the new migration.

It will pass the check in the new code.

This is because the original migration record is already out-of-order

Reply:
On step 2, largestSequence = 1, list[0].Version = "v1.1.2", m.Version = "v1.1.0", a MigrationOutOfOrder error will be returned.

However, yes, we will continue the execution when order-of-ordering exists. This will not happen theoretically except for bugs. If that really happens, it might be okay to continue the execution and ignore the pre-existing ordering because there's no way to fix the existing records and we have to move on.

Reply:
The original logic is more defensive because v1.1.1 will be rejected, but v1.1.3 will work. So we are limiting the damage into that 2 records, instead of bleeding into the new migration record.

Reply:
It would require to search all migration history records (query-wise) to find such out-of-ordering. With the current logic, out-of-ordering could only happen if there's an erratic bug.

Thinking more of the source of out-of-ordering, it comes from that we allow concurrent issues updating the same database. Say v1.1.0 and v1.2.0. The users intend to apply them in order. However, if they click v1.2.0, v1.1.0 cannot be applied anymore.

Reply:
I added a new way of CheckOutOfOrderVersion which should keep the previous behavior.